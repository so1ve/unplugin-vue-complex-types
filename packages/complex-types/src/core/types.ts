import type MagicString from "magic-string";

import type { Printer } from "./printer";

export type ValidTransforms = "defineEmits" | "defineProps";

export type Options = {
	tsconfigPath?: string;
} & {
	[Transform in ValidTransforms]?: boolean;
};

export type ResolvedOptions = Required<Options>;
export type TransformOptions = Pick<ResolvedOptions, ValidTransforms>;

export type Transformer = (
	printer: Printer,
	s: MagicString,
	id: string,
) => void;
export type Transformers = [ValidTransforms, Transformer][];
