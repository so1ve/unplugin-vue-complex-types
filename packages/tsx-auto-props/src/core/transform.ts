import { getLanguage } from "@vue.ts/language";
import MagicString from "magic-string";
import ts from "typescript";
import type { TransformResult } from "unplugin";

const generateDefineProps = (name: string, props: string[]) =>
	`\nObject.defineProperty(${name}, "props", {
  value: ${JSON.stringify(props)},
});`;

export function transform(code: string, id: string): TransformResult {
	const s = new MagicString(code);
	const language = getLanguage();
	const typeChecker = language.__internal__.typeChecker;

	language.traverseAst(id, (node, { ast, isVirtualOrTsFile }) => {
		if (ts.isCallExpression(node)) {
			const identifier = node.expression;
			const symbol = typeChecker.getSymbolAtLocation(identifier);
			if (
				symbol?.declarations?.some(
					(d) =>
						ts.isImportSpecifier(d) &&
						(d.propertyName ?? d.name)?.escapedText === "defineComponent",
				)
			) {
			}
		}
	});

	return {
		code: s.toString(),
		map: s.generateMap({ hires: true }),
	};
}
