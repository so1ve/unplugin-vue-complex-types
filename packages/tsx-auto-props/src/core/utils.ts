import { join } from "node:path";

import type { Options, ResolvedOptions } from "./types";

export const resolveOptions = (rawOptions: Options): ResolvedOptions => ({
	include: rawOptions.include ?? ["**/*.vue", "**/*.tsx"],
	exclude: rawOptions.exclude ?? ["node_modules/**"],
	tsconfigPath: rawOptions.tsconfigPath ?? join(process.cwd(), "tsconfig.json"),
});
