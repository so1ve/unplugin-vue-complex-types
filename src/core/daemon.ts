export type TransformDaemon = {};

export function ensureDaemon() {
	if (!(globalThis as any).__UNPLUGIN_VUE_COMPLEX_DAEMON__) {
		//TODO
		(globalThis as any).__UNPLUGIN_VUE_COMPLEX_DAEMON__ = {};
	}
}

export function getDaemon() {
	ensureDaemon();
	return (globalThis as any).__UNPLUGIN_VUE_COMPLEX_DAEMON__;
}
