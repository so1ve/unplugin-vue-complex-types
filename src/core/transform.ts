import MagicString from "magic-string";
import type { TransformResult } from "unplugin";

import { getLanguage } from "./language";
import { Printer } from "./printer";

export function transform(code: string, id: string): TransformResult {
	const language = getLanguage();
	const s = new MagicString(code);
	const typeChecker = language.__internal__.typeChecker;
	const printer = new Printer(typeChecker);
	const definePropsTypeArg = language.findScriptRangesNode(
		id,
		(scriptSetupRanges) => scriptSetupRanges.props.define?.typeArg,
	);

	if (!definePropsTypeArg) {
		return;
	}

	const {
		node: typeArgNode,
		type: typeArgType,
		range: typeRange,
	} = definePropsTypeArg;

	const printedType = printer.print(typeArgNode);

	s.overwrite(typeRange.start, typeRange.end, printedType);

	return {
		code: s.toString(),
		map: s.generateMap({ hires: true }),
	};
}
