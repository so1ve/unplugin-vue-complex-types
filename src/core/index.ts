import { createUnplugin } from "unplugin";

import { transform } from "./transform";
import type { Options } from "./types";

export default createUnplugin<Options | undefined>((_options) => ({
	name: "unplugin-vue-complex-types",
	transform(code, id) {
		if (!id.endsWith(".vue")) {
			return;
		}

		return transform(code, id);
	},
}));
