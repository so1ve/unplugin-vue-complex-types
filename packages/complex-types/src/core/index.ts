import { createFilter } from "@vue.ts/common";
import { ensureLanguage } from "@vue.ts/language";
import { createUnplugin } from "unplugin";

import { transform } from "./transform";
import type { Options } from "./types";
import { resolveOptions } from "./utils";

export default createUnplugin<Options | undefined>((options = {}) => ({
	name: "@vue.ts/complex-types",
	enforce: 'pre',
	buildStart() {
		const resolvedOptions = resolveOptions(options);
		ensureLanguage(resolvedOptions.tsconfigPath);
	},
	transform(code, id) {
		const file = id.replace(/\?v=.*$/, '');
		const resolvedOptions = resolveOptions(options);
		const filter = createFilter(
			resolvedOptions.include,
			resolvedOptions.exclude,
		);

		if (!filter(file)) {
			return;
		}

		return transform(code, file, resolvedOptions);
	},
}));
