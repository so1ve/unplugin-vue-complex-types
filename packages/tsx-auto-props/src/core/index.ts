import { createFilter } from "@vue.ts/common";
import { ensureLanguage } from "@vue.ts/language";
import { createUnplugin } from "unplugin";

import { transform } from "./transform";
import type { Options } from "./types";
import { resolveOptions } from "./utils";

export default createUnplugin<Options | undefined>((options = {}) => ({
	name: "unplugin-vue-complex-types",
	buildStart() {
		const resolvedOptions = resolveOptions(options);
		ensureLanguage(resolvedOptions.tsconfigPath);
	},
	transform(code, id) {
		const resolvedOptions = resolveOptions(options);
		const filter = createFilter(
			resolvedOptions.include,
			resolvedOptions.exclude,
		);

		if (!filter(id)) {
			return;
		}

		return transform(code, id);
	},
}));
