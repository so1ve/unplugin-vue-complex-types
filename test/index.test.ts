import { join } from "node:path";

import { describe, expect, it } from "vitest";

import { transform } from "../src/core/transform";
import AVue from "./__fixtures__/a.vue?raw";

describe("should", () => {
	it("exported", () => {
		expect(
			transform(AVue, join(__dirname, "__fixtures__", "a.vue")),
		).toMatchSnapshot();
	});
});
