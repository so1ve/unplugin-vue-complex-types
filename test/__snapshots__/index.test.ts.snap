// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should > exported 1`] = `
{
  "code": "<script setup lang=\\"ts\\">
import type { A } from \\"./foo\\";

defineProps<
	{
  msg: string;
  aaa?: string;
  bbb: string;
  foo: object;
  bar: number;
  baz: object;
}
>();
</script>
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACxB,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAC/B;AACA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACZ,CAAC;;;;;;;CAIK;AACN,CAAC,CAAC,CAAC,CAAC;AACJ,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
